src/plugins/intel_gpu/src/graph/loop.cpp

```diff
@@ -316,181 +381,275 @@ void loop_inst::update_mapped_memory() {
-            concatenated_output_mem_mappings.push_back(memory_mapping_info);
         }
     }
 }
 
-void loop_inst::preprocess_input_memory() {
+void loop_inst::preprocess_input_memory(const int64_t num_iterations) {
     auto& engine = _network.get_engine();
     auto& iteration_mem = concatenated_input_mem_mappings;
     for (size_t memory_num = 0; memory_num < inputs_memory_count(); memory_num++) {
         const primitive_id& input_external_id = dependencies().at(memory_num).first->id();
         auto input_map_ptrs = find_io_primitive_maps(_input_primitive_maps,
-                                                     _output_primitive_maps, input_external_id, true);
+                                                    _output_primitive_maps, input_external_id, true);
         if (input_map_ptrs.size() == 0) {
-            if (input_external_id == _trip_count_id ||
-                input_external_id == _initial_execution_id) {
-                continue;
-            }
-            CLDNN_ERROR_MESSAGE(id(), "loop primitive_map is incomplete");
+            OPENVINO_ASSERT((input_external_id == _trip_count_id
+                                || input_external_id == _num_iterations_id
+                                || input_external_id == _initial_execution_id),
+                                id(), "loop primitive_map is incomplete "
+                                "input_external_id(", input_external_id, ") != _trip_count_id(", _trip_count_id, ")",
+                                "input_external_id(", input_external_id, ") != _num_iterations_id(", _num_iterations_id, ")",
+                                " && input_external_id(", input_external_id, ") != _initial_execution_id(", _initial_execution_id, ")");
+            continue;
         }
 
         auto memory = input_memory_ptr(memory_num);
         for (size_t i = 0; i < input_map_ptrs.size(); ++i) {
             const auto input_map = input_map_ptrs.at(i);
+            const auto& external_id = input_map->external_id;
+            const auto& internal_id = input_map->internal_id;
+            GPU_DEBUG_LOG << i << ") input mapping - external " << external_id.to_string() << std::endl;
+            GPU_DEBUG_LOG << i << ") input mapping - internal " << internal_id.to_string() << std::endl;
+
             bool is_concatenated_input = (input_map->axis >= 0);
             if (is_concatenated_input) {
                 layout sliced_layout
-                    = body_network->get_primitive(input_map->internal_id)->output_memory().get_layout();
-                const int64_t max_iteration = _max_iteration;
+                    = body_network->get_primitive(internal_id.pid)->output_memory(internal_id.idx).get_layout();
                 std::vector<memory::ptr> sliced_mems;
-                sliced_mems.reserve(max_iteration);
-                for (int j=0; j < max_iteration; ++j) {
+                if (num_iterations < 0) {
```

Comment:
I think, if "concatenated input" is used, then the # iterations are already decided at the time of execution, e.g., same as # of sliced mem.s

Reply:
Yes, You're right.  In preprocess_input_memory, num_iterations should be positive.  Actually I'm going to add this checking code in preprocess_input_memory but I was confused so that I added it to preprocess_output_memory. 
Now, I fixed codes. please review it again.